### ISSUE TITLE
Poor real variable mixing (Add HMC moves for real variables) 

### ISSUE THREAD

**Issue by jewellsean** [alexandrebouchard/bayonet#28]
I've been prototyping a few simple examples in Blang based on Stat 520 (see, for example, [here](http://www.stat.ubc.ca/~bouchard/courses/stat520-sp2014-15/homework/2015/02/23/assignment1.html)) and it seems that there is poor mixing. 

One reason is that only one move is being used for real variables (the Peskun move is failing similarly to #6). I plan to address #6 shortly, but also want to solicit feedback on including the AHMC library's HMC functionality as another move for real variables. 

Any feedback is appreciated.

**Comment by alexandrebouchard** [comment]
Sounds good. Do you mean using the adaptive BO for MH hyperparameter tuning, or with the actual HMC as well? In the latter case, the issue is that we would need automatic differentiation. I guess we could resort to numerical differentiation too. What option did you have in mind?

**Comment by jewellsean** [comment]
Thanks for the comment, Alex. I had been thinking along the lines of HMC + numerical differentiation (and perhaps doing a block HMC move on all real variables -- I have been working on IID collections but this could be generalized).

I took a quick look at BO for the MH Hyperparameter tuning as discussed in [Mahendran et al 2012](http://www.cs.ubc.ca/~ziyuw/papers/mahendran12.pdf). This might be an easier approach? I am not very familiar with the AHMC library so I am not confident that it can help too much here, but the approach seems to address issues we had in tuning the ASP MH hyperparameters (which was tedious to do manually). Perhaps you are more familiar with Ziyu's code? I am willing to develop this feature into the core blang moveset.

**Comment by alexandrebouchard** [comment]
An issue with HMC + numerical is that each numerical diff will take n evaluations, where n is the number of real variables (since we need all n partial derivatives). But I guess we could do HMC within block Gibbs to address this, taking blocks to be variables that are connected and therefore perhaps more likely to be correlated.

Alternatively, something that might be easier to implement is slice sampling.

**Comment by jewellsean** [comment]
Thanks! Due to time constraints, I will first try a slice implementation (it seems straightforward to implement). Builds reflecting this change are forthcoming (next few days).

**Comment by jewellsean** [comment]
Summary of changes before I close this issue: 

(i) Created a new branch of bayonet called ``dev``. As the name suggests, we can put in-development items in here until they have passed all of the tests and have been vetted. I've committed a basic slice sampler here. 

(ii) I've added the overrelaxed slice sampler of Neal 2003 to the main branch. It has been added to the list of samplers for RealVariables. In my testing it seems to improve mixing. 

(iii) In the distribution unit tests, it is not entirely clear why some samplers *work* while others do not. It seems we need some sort of policy or set of units tests that all samplers should pass.
